### PR TITLE
[compiler v2] Integration into package system and Aptos CLI

### PR DESCRIPTION
**Pull request by wrwg**
### Description

This integrates the v2 compiler into the `move-package` code and makes it available to the `aptos` CLI. One can now compile with the v2 compiler as in

```
aptos move compile --compiler-version=v2
```

However, compilation of already the Move stdlib fails because of #9330 (effecting bit operations in `fixed_point32.move`) and #9696 (effecting inline functions with lambdas in `vector.move`). Fixing those issues does not fit in this PR which is by itself not trivial.

Implementation wise, since the package system is build on top of internals of the v1 compiler, we provide a new v2 driver function parameter to the general entrypoint `compile_with_driver`. We also add a new flag for the compiler version which is plumbed through many layers of configurations from Aptos CLI down to Move compiler (something which cleaning up is out of scope for this PR).

This PR also disables the use of the bytecode produced by v2 for production, by assigning a new experimental version number to it which is only allowed in `testing` feature mode.

### Test Plan

Tested manually.

### CODE REVIEW

File: third_party/move/move-binary-format/src/file_format_common.rs [hunk 1]
```diff
@@ -405,9 +405,14 @@ pub const VERSION_5: u32 = 5;
...
 pub const VERSION_6: u32 = 6;
 
-// Mark which version is the latest version
+/// Mark which version is the latest version
 pub const VERSION_MAX: u32 = VERSION_6;
 
+/// A unique version value which is used for experimental code which is not allowed in
+/// production. The bytecode deserializer accepts modules with this version only when the
+/// cargo feature `testing` is enabled.
+pub const VERSION_EXPERIMENTAL: u32 = u32::MAX;
```
Comment (rahxephon89) [inline third_party/move/move-binary-format/src/file_format_common.rs:414]: We will keep this until compiler v2 is in production?

File: third_party/move/tools/move-package/src/lib.rs [hunk 2]
```diff
@@ -148,6 +154,22 @@ pub struct BuildConfig {
...
+    /// Compiler version to use
+    #[clap(long = "compiler-version", global = true)]
+    pub compiler_version: Option<CompilerVersion>,
+}
+
+#[derive(ValueEnum, Clone, Copy, Debug, Serialize, Deserialize, Eq, PartialEq, PartialOrd)]
+pub enum CompilerVersion {
+    V1,
+    V2,
+}
```
Comment (gregnazario) [inline third_party/move/tools/move-package/src/lib.rs:167]: We may want to consider more fine grained compiler versions in the future.

File: third_party/move/tools/move-cli/src/base/test.rs [hunk 3]
```diff
@@ -13,13 +13,14 @@ use move_compiler::{
     PASS_CFGIR,
 };
 use move_coverage::coverage_map::{output_map_to_file, CoverageMap};
-use move_package::{compilation::build_plan::BuildPlan, BuildConfig};
+// if windows
```
Comment (gregnazario) [inline third_party/move/tools/move-cli/src/base/test.rs:16]: This looks like an awkward place to have this, is it supposed to be conditional inputs?

File: third_party/move/testing-infra/transactional-test-runner/src/framework.rs [hunk 4]
```diff
@@ -680,18 +680,15 @@ fn compile_source_unit_v2(
...
     let error_str = String::from_utf8_lossy(&error_writer.into_inner()).to_string();
-    let (mut modules, mut scripts) =
+    let (_, mut units) =
         result.map_err(|_| anyhow::anyhow!("compilation errors:\n {}", error_str))?;
-    let unit = if modules.is_empty() && scripts.len() == 1 {
-        (None, scripts.pop())
-    } else if scripts.is_empty() && modules.len() == 1 {
-        (modules.pop(), None)
+    let unit = if units.len() != 1 {
+        anyhow::bail!("expected either one script or one module")
```
Comment (gregnazario) [inline third_party/move/testing-infra/transactional-test-runner/src/framework.rs:686]: Any reason not to give the module and scripts size anymore?

File: third_party/move/move-compiler-v2/transactional-tests/tests/control_flow/sorter.exp [hunk 5]
```diff
@@ -286,7 +286,7 @@ B0:
 
 >>> V2 Compiler {
 == BEGIN Bytecode ==
-// Move bytecode v6
+// Move bytecode v4294967295
```
Comment (gregnazario) [inline third_party/move/move-compiler-v2/transactional-tests/tests/control_flow/sorter.exp:289]: This seems wrong

File: third_party/move/move-compiler-v2/transactional-tests/tests/control_flow/sorter.exp [hunk 6]
```diff
@@ -567,7 +567,7 @@ B5:
 task 1 'run'. lines 67-73:
 
 == BEGIN Bytecode ==
-// Move bytecode v6
+// Move bytecode v4294967295
```
Comment (gregnazario) [inline third_party/move/move-compiler-v2/transactional-tests/tests/control_flow/sorter.exp:570]: Similarly

File: third_party/move/move-compiler-v2/tests/file-format-generator/assign.exp [hunk 7]
```diff
@@ -106,7 +106,7 @@ fun assign::assign_struct($t0: &mut assign::S) {
 
 
 ============ disassembled file-format ==================
-// Move bytecode v6
```
Comment (gregnazario) [inline third_party/move/move-compiler-v2/tests/file-format-generator/assign.exp:109]: Similarly on the move bytecode version needs to be checked

File: third_party/move/move-compiler/src/compiled_unit.rs [hunk 8]
```diff
@@ -44,8 +44,6 @@ pub struct SpecInfo {
 #[derive(Debug, Clone)]
 pub struct FunctionInfo {
     pub spec_info: BTreeMap<SpecId, SpecInfo>,
-    pub parameters: Vec<(Var, VarInfo)>,
-    pub attributes: Attributes,
 }
```
Comment (gregnazario) [inline third_party/move/move-compiler/src/compiled_unit.rs:47]: Where is this stuff stored now?

File: third_party/move/move-analyzer/src/symbols.rs [hunk 9]
```diff
@@ -645,40 +648,45 @@ impl Symbolicator {
...
+                    Err(diags) => {
+                        let failure = false;
+                        diagnostics = Some((diags, failure));
+                        eprintln!("bytecode compilation failed");
+                        return Ok((files, vec![]));
+                    },
+                };
+                // warning diagnostics (if any) since compilation succeeded
+                if !diags.is_empty() {
+                    // assign only if non-empty, otherwise return None to reset previous diagnostics
```
Comment (gregnazario) [inline third_party/move/move-analyzer/src/symbols.rs:681]: Please move this to a separate function for readability purposes.

File: third_party/move/move-binary-format/src/file_format_common.rs (hunk 1 above)
Comment (gregnazario) [inline third_party/move/move-binary-format/src/file_format_common.rs:414]: Is there supposed to be a different version of the bytecode for the V2 compiler, or will it be completely compatible with the old version?

File: third_party/move/move-compiler-v2/Cargo.toml [hunk 10]
```diff
@@ -12,10 +12,15 @@ edition = "2021"
 [dependencies]
 anyhow = "1.0.62"
 move-binary-format = { path = "../move-binary-format" }
+move-bytecode-source-map = { path = "../move-ir-compiler/move-bytecode-source-map" }
+move-command-line-common = { path = "../move-command-line-common" }
+#move-package = { path = "../tools/move-package" }
```
Comment (gregnazario) [inline third_party/move/move-compiler-v2/Cargo.toml:17]: let's remove this line

File: third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs [hunk 11]
```diff
@@ -83,12 +90,26 @@ impl ModuleGenerator {
...
+                let name = IR_AST::ModuleName(IR_SYMBOL::Symbol::from(
+                    ctx.symbol_to_str(module_env.get_name().name()),
+                ));
+                Some(IR_AST::ModuleIdent::new(
+                    name,
+                    module_env.get_name().addr().expect_numerical(),
+                ))
+            };
+            SourceMap::new(ctx.env.to_ir_loc(&module_env.get_loc()), module_name_opt)
+        };
```
Comment (gregnazario) [inline third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs:112]: Any particular reason it's in the surrounding block?  Makes it a little less readable

**Review by gregnazario — APPROVED** [review]
Seems mostly fine, I mostly have readability comments.

The UNUSED one, maybe we should make just a really high known number, since I suspect we'll never get that high, and it'll be more obvious, like 60,000

File: third_party/move/move-binary-format/src/file_format_common.rs (hunk 1 above)
Comment (wrwg) [reply third_party/move/move-binary-format/src/file_format_common.rs:414]: In the first iteration, we aim at v6 (the current version), so fully compatible.

File: third_party/move/move-compiler-v2/Cargo.toml (hunk 10 above)
Comment (wrwg) [reply third_party/move/move-compiler-v2/Cargo.toml:17]: Will do next iteration

File: third_party/move/move-analyzer/src/symbols.rs (hunk 9 above)
Comment (wrwg) [reply third_party/move/move-analyzer/src/symbols.rs:681]: That's actually old code, it was just indented. I agree it looks very messy. Hopefully we retire this path together with the old compiler. I factor it out next time I get along.

File: third_party/move/tools/move-cli/src/base/test.rs (hunk 3 above)
Comment (wrwg) [reply third_party/move/tools/move-cli/src/base/test.rs:16]: This was moved by autoformatter after reformating imports. Move it back to where it was, before the cfg!

**Pull request merged**